File: services/api/src/aec_api/routers/client_portal.py
"""CLIENT-PORTAL routes — share-token management (owner-authenticated) + the public read-only digest.

The management routes are gated to editors on the project; the public digest route takes only the token
(the token IS the credential) and exposes a curated readiness summary. An unknown/revoked token 404s.
"""
from __future__ import annotations

from fastapi import APIRouter, Body, Depends, HTTPException, Response
from sqlalchemy.orm import Session  # noqa: F401

from .. import client_portal
from ..db import get_db
from ..models import Project
from ..rbac import require_role

router = APIRouter()


@router.post("/projects/{pid}/share-tokens")
def create_share_token(pid: str, body: dict = Body(default={}), db: Session = Depends(get_db),
                       actor: str = Depends(require_role("editor"))):
    """Mint a revocable, read-only share token for the project (owner/editor only). The returned token
    string is the shareable credential — anyone with it can read the project's readiness digest.
    `show_payments: true` is the explicit OPT-IN for THIS token's digest to carry the owner-invoice
    payment schedule (display only); the default digest exposes no financials."""
    if not db.get(Project, pid):
        raise HTTPException(404, "project not found")
    try:
        return client_portal.create_token(db, pid, body.get("label"), actor,
                                          show_payments=bool(body.get("show_payments")))
    except ValueError as e:
        raise HTTPException(409, str(e)) from None


@router.get("/projects/{pid}/share-tokens")
def list_share_tokens(pid: str, db: Session = Depends(get_db), _: str = Depends(require_role("editor"))):
    """List the project's share tokens with their view counts (editor only)."""
    if not db.get(Project, pid):
        raise HTTPException(404, "project not found")
    return {"tokens": client_portal.list_tokens(db, pid)}


@router.delete("/projects/{pid}/share-tokens/{token}")
def revoke_share_token(pid: str, token: str, db: Session = Depends(get_db),
                       _: str = Depends(require_role("editor"))):
    """Revoke a share token — access stops immediately (editor only)."""
    if not db.get(Project, pid):
        raise HTTPException(404, "project not found")
    if not client_portal.revoke(db, pid, token):
        raise HTTPException(404, "token not found or already revoked")
    return {"revoked": True}


@router.get("/projects/{pid}/client-decisions")
def list_client_decisions(pid: str, limit: int = 500, db: Session = Depends(get_db),
                          _: str = Depends(require_role("editor"))):
    """PORTAL-TXN — the project's client-decision feed (approve/acknowledge/decline recorded through share
    tokens), newest first (editor only)."""
    if not db.get(Project, pid):
        raise HTTPException(404, "project not found")
    return {"decisions": client_portal.decisions_for_project(db, pid, limit)}


@router.post("/shared/{token}/decision")
def shared_decision(token: str, body: dict = Body(...), db: Session = Depends(get_db)):
    """PUBLIC (no auth) — PORTAL-TXN: record a client decision through a live share token — a timestamped,
    token-stamped **approve / acknowledge / decline** on a shared item. NOT a payment and NOT an e-signature
    of record. Inputs are whitelisted + length-capped and each token carries a hard decision cap. Body:
    `{item_type, item_ref, action, client_name?, note?}`. Unknown/revoked token → 404."""
    try:
        return client_portal.record_decision(db, token, body.get("item_type"), body.get("item_ref"),
                                             body.get("action"), body.get("client_name"), body.get("note"))
    except KeyError:
        raise HTTPException(404, "not found") from None
    except ValueError as e:
        raise HTTPException(409 if "limit" in str(e) else 422, str(e)) from None


@router.get("/shared/{token}/digest")
def shared_digest(token: str, db: Session = Depends(get_db)):
    """PUBLIC (no auth) — the curated read-only project digest for a valid share token. High-level
    readiness only; no record-level data. An unknown or revoked token returns 404 (no enumeration)."""
    try:
        return client_portal.digest(db, token)
    except KeyError:
        raise HTTPException(404, "not found") from None


@router.get("/shared/{token}")
def shared_page(token: str, db: Session = Depends(get_db)):
    """PUBLIC (no auth) — a self-contained read-only HTML page rendering the share digest (the same
    curated readiness data as the .json digest). All values are HTML-escaped. Unknown/revoked → 404."""
    try:
        d = client_portal.digest(db, token)
    except KeyError:
        raise HTTPException(404, "not found") from None
    return Response(client_portal.to_html(d), media_type="text/html")

File: apps/web/src/api/clientPortal.ts
/** Client portal — how does someone with no account see this project, and answer back?
 *
 *  SCALE-SEAM (90), third slice from the methods above the STAYING banner.
 *
 *  Two halves of one question. The OWNER's half mints, lists and revokes share tokens
 *  (`createShareToken`, `shareTokens`, `revokeShareToken`) and reads what came back
 *  (`clientDecisions`). The RECIPIENT's half is the token-authenticated public surface:
 *  `sharedPageUrl` and `sharedDigestUrl` build the no-login URLs, and `sharedComment` and
 *  `sharedDecision` post back a comment or an approve/acknowledge/decline.
 *
 *  They belong together because the token is the seam: it is minted on one side, IS the credential
 *  on the other, and revoking it closes both. Splitting owner-side from recipient-side would put a
 *  capability and its only means of exercise in different files.
 *
 *  `services/api/src/aec_api/routers/client_portal.py` groups the same set — 8 of its 9 routes are
 *  exactly these methods, checked rather than assumed. The ninth, `GET /shared/{token}/model.frag`,
 *  still has no client method: it is a whole-file response a recipient opens by URL, not JSON this
 *  SPA parses. What it DOES need from here is the mint-time opt-in that makes it answer at all.
 *
 *  ### The opt-in this file could not reach — fixed 2026-09-04
 *
 *  That ninth route serves geometry only to a token minted with `show_model`, which the backend
 *  treats as an opt-in independent of `show_payments` — *"a token may carry payments, or geometry,
 *  or neither, and granting one never implies the other"*. **`createShareToken` did not send it**,
 *  and the row type `shareTokens` returns had no `show_model` field, so every token this product
 *  minted had it false and that route 404'd for all of them.
 *
 *  Both halves are closed below, and they are separate defects with separate consequences. The
 *  missing PARAMETER made the capability unreachable. The missing ROW FIELD made it unauditable —
 *  `_public_row` has always returned `show_model`, the wire carried it, and the type simply dropped
 *  it, so an owner could not have told a geometry link from a digest link even once one existed.
 *  R22-PUBLIC-VIEWER's shipped record claims *"the owner's token list shows which links carry
 *  geometry"*; until this change that was true of the JSON and false of the product.
 *
 *  The two flags are passed as separate arguments, never as one "share more" level, because the
 *  backend's rule is that granting one must never imply the other. A single enum or an ordered
 *  level would make that rule unexpressible at the call site.
 *
 *  A mixin, so every call site resolves unchanged; `api/surface.test.ts` is what proves it.
 */
import { HttpCore } from "./httpCore";

type Ctor<T> = new (...args: any[]) => T;

export function withClientPortal<TBase extends Ctor<HttpCore>>(Base: TBase) {
  return class ClientPortal extends Base {
  /** Mint a read-only share token. `showPayments` and `showModel` are two INDEPENDENT opt-ins, and
   * the backend is explicit that granting one never implies the other: `showPayments` lets this
   * token's digest carry the owner-invoice payment schedule, `showModel` lets it fetch the project's
   * geometry fragment (`GET /shared/{token}/model.frag` — shapes and placements, never the source
   * IFC). Both default to false; a token already in somebody's inbox is never widened. */
  createShareToken(pid: string, label?: string, showPayments?: boolean, showModel?: boolean) {
    return this.json<{ token: string; label: string | null; share_path: string; revoked: boolean;
      show_payments: boolean; show_model: boolean }>(
      `/projects/${pid}/share-tokens`,
      { method: "POST", body: JSON.stringify({ label: label ?? "", show_payments: !!showPayments,
        show_model: !!showModel }) });
  }
  /** CLIENT-PORTAL — read-only share tokens for a project-readiness digest. */
  shareTokens(pid: string) {
    type Tok = { token: string; label: string | null; revoked: boolean; created_at: string | null;
      created_by: string | null; view_count: number; last_viewed_at: string | null; share_path: string;
      show_payments: boolean; show_model: boolean };
    return this.json<{ tokens: Tok[] }>(`/projects/${pid}/share-tokens`);
  }
  revokeShareToken(pid: string, token: string) {
    return this.json<{ revoked: boolean }>(`/projects/${pid}/share-tokens/${encodeURIComponent(token)}`,
      { method: "DELETE" });
  }
  /** The public read-only HTML page for a share token (opens with no login — the human share link). */
  sharedPageUrl(token: string) { return this.url(`/shared/${encodeURIComponent(token)}`); }
  /** The public digest JSON URL for a share token. */
  sharedDigestUrl(token: string) { return this.url(`/shared/${encodeURIComponent(token)}/digest`); }
  /** PORTAL-TXN phase 3 — post a client comment through a share token (public; lands on the token's
   * BCF feedback topic, so the team answers from the Issue Board). */
  sharedComment(token: string, body: { text: string; client_name?: string }) {
    return this.json<{ topic_id: string; comment_id: string; author: string | null; text: string;
      created_at: string | null }>(
      `/shared/${encodeURIComponent(token)}/comment`, { method: "POST", body: JSON.stringify(body) });
  }
  /** PORTAL-TXN — record a client decision through a share token (public; approve/acknowledge/decline). */
  sharedDecision(token: string, body: {
    item_type: string; item_ref: string; action: "approved" | "acknowledged" | "declined";
    client_name?: string; note?: string;
  }) {
    return this.json<{
      id: number; item_type: string; item_ref: string; action: string;
      client_name: string | null; note: string | null; created_at: string | null;
    }>(`/shared/${encodeURIComponent(token)}/decision`, { method: "POST", body: JSON.stringify(body) });
  }
  /** PORTAL-TXN — the project's client-decision feed (editor only), newest first. */
  clientDecisions(pid: string, limit = 500) {
    type D = { id: number; item_type: string; item_ref: string; action: string; client_name: string | null;
      note: string | null; created_at: string | null; token: string };
    return this.json<{ decisions: D[] }>(`/projects/${pid}/client-decisions?limit=${encodeURIComponent(limit)}`);
  }
  };
}
